File: examples/smoke-python-search/bench.py
import time

from search import build_demo_catalog, demo_queries, search_catalog


CATALOG = build_demo_catalog(320)
QUERIES = demo_queries() * 6
SAMPLES = 15


def measure_once():
    started = time.perf_counter()
    for query in QUERIES:
        search_catalog(CATALOG, query, limit=5)
    elapsed = time.perf_counter() - started
    return elapsed * 1000.0


def percentile_95(samples):
    ordered = sorted(samples)
    index = round((len(ordered) - 1) * 0.95)
    return ordered[index]


def main():
    samples = [measure_once() for _ in range(SAMPLES)]
    print(f"METRIC latency_p95={percentile_95(samples):.3f}")


if __name__ == "__main__":
    main()

File: examples/smoke-rust-cli/src/bin/bench.rs
use std::time::Instant;

use smoke_rust_cli::{demo_catalog, suggest_commands};

const SAMPLES: usize = 15;

fn percentile_95(samples: &mut [f64]) -> f64 {
    samples.sort_by(|left, right| left.partial_cmp(right).unwrap());
    let index = ((samples.len() - 1) as f64 * 0.95).round() as usize;
    samples[index]
}

fn measure_once() -> f64 {
    let catalog = demo_catalog(320);
    let queries = [
        "cache memoization performance",
        "bench latency metrics",
        "workspace diagnostics",
        "log triage errors",
        "reproducible benchmark setup",
        "performance cache latency",
    ];

    let started = Instant::now();
    for _ in 0..6 {
        for query in queries {
            let _ = suggest_commands(&catalog, query, 5);
        }
    }
    started.elapsed().as_secs_f64() * 1000.0
}

fn main() {
    let mut samples = (0..SAMPLES).map(|_| measure_once()).collect::<Vec<_>>();
    println!("METRIC latency_p95={:.3}", percentile_95(&mut samples));
}
